### DISCUSSION TITLE
`Query` default value cannot be set in `Annotated` when using `APIRouter`

### DISCUSSION THREAD

**Discussion by tlinhart** [fastapi/fastapi#9279] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# main.py

from fastapi import FastAPI
from router import router

app = FastAPI()
app.include_router(router)

# router.py

from typing import Annotated
from fastapi import APIRouter, Query

router = APIRouter()

@router.get("/")
async def root(*, n: Annotated[float, Query()] = 50):
    return {"n": n}
```

### Description

When starting the Uvicorn server using `uvicorn main:app`, I receive:

```
Traceback (most recent call last):
  File "/home/pasmen/tmp/x-annotated-query-router/./venv/bin/uvicorn", line 8, in <module>
    sys.exit(main())
  File "/home/pasmen/tmp/x-annotated-query-router/venv/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/home/pasmen/tmp/x-annotated-query-router/venv/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/home/pasmen/tmp/x-annotated-query-router/venv/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/pasmen/tmp/x-annotated-query-router/venv/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/home/pasmen/tmp/x-annotated-query-router/venv/lib/python3.10/site-packages/uvicorn/main.py", line 403, in main
    run(
  File "/home/pasmen/tmp/x-annotated-query-router/venv/lib/python3.10/site-packages/uvicorn/main.py", line 568, in run
    server.run()
  File "/home/pasmen/tmp/x-annotated-query-router/venv/lib/python3.10/site-packages/uvicorn/server.py", line 59, in run
    return asyncio.run(self.serve(sockets=sockets))
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "uvloop/loop.pyx", line 1517, in uvloop.loop.Loop.run_until_complete
  File "/home/pasmen/tmp/x-annotated-query-router/venv/lib/python3.10/site-packages/uvicorn/server.py", line 66, in serve
    config.load()
  File "/home/pasmen/tmp/x-annotated-query-router/venv/lib/python3.10/site-packages/uvicorn/config.py", line 471, in load
    self.loaded_app = import_from_string(self.app)
  File "/home/pasmen/tmp/x-annotated-query-router/venv/lib/python3.10/site-packages/uvicorn/importer.py", line 21, in import_from_string
    module = importlib.import_module(module_str)
  File "/usr/lib/python3.10/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/home/pasmen/tmp/x-annotated-query-router/main.py", line 5, in <module>
    app.include_router(router)
  File "/home/pasmen/tmp/x-annotated-query-router/venv/lib/python3.10/site-packages/fastapi/applications.py", line 427, in include_router
    self.router.include_router(
  File "/home/pasmen/tmp/x-annotated-query-router/venv/lib/python3.10/site-packages/fastapi/routing.py", line 782, in include_router
    self.add_api_route(
  File "/home/pasmen/tmp/x-annotated-query-router/venv/lib/python3.10/site-packages/fastapi/routing.py", line 600, in add_api_route
    route = route_class(
  File "/home/pasmen/tmp/x-annotated-query-router/venv/lib/python3.10/site-packages/fastapi/routing.py", line 446, in __init__
    self.dependant = get_dependant(path=self.path_format, call=self.endpoint)
  File "/home/pasmen/tmp/x-annotated-query-router/venv/lib/python3.10/site-packages/fastapi/dependencies/utils.py", line 303, in get_dependant
    type_annotation, depends, param_field = analyze_param(
  File "/home/pasmen/tmp/x-annotated-query-router/venv/lib/python3.10/site-packages/fastapi/dependencies/utils.py", line 387, in analyze_param
    assert field_info.default is Undefined or field_info.default is Required, (
AssertionError: `Query` default value cannot be set in `Annotated` for 'n'. Set the default value with `=` instead.
```

### Operating System

Linux

### Operating System Details

`lsb_release -a`:

```
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 22.04.2 LTS
Release:	22.04
Codename:	jammy
```

### FastAPI Version

0.95.0

### Python Version

Python 3.10.6

### Additional Context

This should work according to the [documentation about default values](https://fastapi.tiangolo.com/tutorial/query-params-str-validations/#default-values). Also, it works when not using `APIRouter` but having the path operation directly on `app`.

**Comment by vargasjona** [comment]
I have the same problem this happens also when the time is date and enum those endpoints are also in a router. I have this error message in both cases for the moment the solution is not using Annotated:

`AssertionError: `Query` default value cannot be set in `Annotated` for`

Working code
```
    user_status: IUserStatus = Query(
        default=IUserStatus.active,
        title= "User status",
        description="User status, It is optional. Default is active",
    ),
```

Failing code
```
 user_status: Annotated[
        IUserStatus, 
        Query(        
            title= "User status",
            description="User status, It is optional. Default is active",
        )
    ] = IUserStatus.active,
```


Working code
```
    start_date: date = Query(default=(datetime.now() - timedelta(days=7)).date()),
    end_date: date = Query(default=datetime.now().date()),
```

Failing code
```
    start_date: Annotated[date, Query(title="start date for get data")] = datetime.now() - timedelta(days=7),
    end_date: Annotated[date, Query(title="end date for get data")] = datetime.now().date(),
```

**Reply by tiangolo** [reply]
This should be fixed in https://github.com/tiangolo/fastapi/pull/9315

Just released in FastAPI 0.95.1 :tada:

**Reply by vargasjona** [reply]
Thanks @tiangolo it works perfectly :)

**Comment by hochstibe** [comment]
I have the same problem, it depends on the ordering: If everything is in one file, the router can be added before defining the endpoint.
If the router is in a separate file (like tlinhart), this is not possible. The same error resulted in Header, Path or Query parameters.

```python
from typing import Annotated
from fastapi import FastAPI, Header, APIRouter

app = FastAPI()

@app.get("/app")
async def read_items(user_agent: Annotated[str | None, Header()] = None):
    return {"User-Agent": user_agent}

test_router = APIRouter()


# if the router is added before defining the route, it works fine
# app.include_router(test_router)

@test_router.get("/router")
async def read_items(user_agent: Annotated[str | None, Header()] = None):
    return {"User-Agent": user_agent}

# breaks
app.include_router(test_router)
```

**Reply by tiangolo** [reply]
This should be fixed in https://github.com/tiangolo/fastapi/pull/9315

Just released in FastAPI 0.95.1 :tada:

**Comment by EricHuang0525** [comment]
I am facing the same problem here. This only happens when endpoints use `APIRouter` but not `FastAPI`.

**Reply by tiangolo** [reply]
This should be fixed in https://github.com/tiangolo/fastapi/pull/9315

Just released in FastAPI 0.95.1 :tada:

**Comment by AlexanderPodorov** [comment]
Just want to put my 2 cents into this topic. Those dependencies don't work either:
```python
# function dependency
def common_parameters(
        q: Annotated[str | None, Query()] = 'asd',
        skip: Annotated[int, Query()] = 4,
        limit: Annotated[int, Query()] = 5
):
    return {"q": q, "skip": skip, "limit": limit}


# class as dependency
class CommonQueryParams:
    def __init__(
            self,
            q: Annotated[str | None, Query()] = None,
            skip: Annotated[int, Query()] = 0,
            limit: Annotated[int, Query()] = 100
    ):
        self.q = q
        self.skip = skip
        self.limit = limit


# dataclass as dependency (in order to get pregenerated __init__)
@dataclass
class CommonQueryParamsDC:
    q: Annotated[str | None, Query()] = None
    skip: Annotated[int, Query()] = 0
    limit: Annotated[int, Query()] = 100
```
The same error message is returned.

**Reply by tiangolo** [reply]
This should be fixed in https://github.com/tiangolo/fastapi/pull/9315

Just released in FastAPI 0.95.1 :tada:

**Comment by AlexanderPodorov** [comment]
Could be unrelated issue, I'm using `pydantic` dataclass for Query dependencies. This way I'm benefiting from automatic aliasing provided by `CustomBaseModel` (alias_generator is defined there). This code works:
```python
@dataclass(config=CustomBaseModel.Config)
class PaginationGet:
    page: int = Query(1, gt=0)
    per_page: PerPage = Query(PerPage.VALUE_025)
    all_items: bool = Query(False)
```
This code does not work:
```python
@dataclass(config=CustomBaseModel.Config)
class PaginationGet:
    page: Annotated[int, Query(gt=0)]
    per_page: Annotated[PerPage, Query()]
    all_items: Annotated[bool, Query()]
```
The error is this:
`ValueError: cannot specify Annotated and value Fields together for 'page'`

**Reply by tiangolo** [reply]
Hum, this is actually not currently supported in FastAPI, if it works it's a bit by "accident" :sweat_smile: 

But I want to have a way to support things like this in the future. :nerd_face:

**Comment by vargasjona** [comment]
Why this is in discussions when it seems a bug on the apirouter of latest fastapi ?

**Reply by tlinhart** [reply]
> Why this is in discussions when it seems a bug on the apirouter of latest fastapi ?

Mostly because of this https://twitter.com/tiangolo/status/1632046370021425154.

**Reply by jack-michaud** [reply]
When I searched the issues evaluating the upgrade to FastAPI 0.95.0, I saw a small amount of issues and none of them bugs with the new version. If I saw this thread, I would have re-evaluated since my large project _does_ use APIRouters... It would be helpful to upgrade discussions to Issues more proactively.

**Reply by tlinhart** [reply]
@jack-michaud agreed. I might be wrong but there seem to be no guidelines for reporting issues. I've gone through https://fastapi.tiangolo.com/contributing but it doesn't talk about reporting bugs. Seeing how most of the issues were transferred into discussions, I thought that would be the way to go.

**Reply by hochstibe** [reply]
I see the point with only allowing discussions. But then the bug reports should be regularly converted as issues....

**Reply by leokhachatorians** [reply]
This entire concept of blanket migrating _all_ issues into discussions is ridiculous. It's literally sweeping crap under the couch and pretending your living room is clean.

**Reply by tiangolo** [reply]
Thanks for the feedback.

Yep, here's the issue: https://github.com/tiangolo/fastapi/issues/9387

@leokhachatorians I would appreciate if you tried to be a bit more gentle with your words, you're getting open source free software, for free, please, at least try to be kind.

---

For more context to all, before migrating to Discussions, around 90% of the "issues" were questions. And before that, when it was possible to report "bugs", 90% of the reported "bugs" were actually user errors. So, unfortunately, I can't let the space open to freely create issues and/or mark them as bugs to everyone.

Answering questions (here or in issues) is what consumes most of the time, so I have to try and structure it the best way I can and try to reduce the noise, to also be able to work a bit on the code, docs, etc.

If any of you has the chance, your help by helping others with their questions here in Discussions would be greatly appreciated: https://fastapi.tiangolo.com/help-fastapi/#help-others-with-questions-in-github

**Comment by me-v2** [comment]
this is error issues. @tiangolo

**Reply by tiangolo** [reply]
This should be fixed in https://github.com/tiangolo/fastapi/pull/9315

Just released in FastAPI 0.95.1 :tada:

**Comment by sharonyogev** [comment, marked as answer]
This should solve it: https://github.com/tiangolo/fastapi/pull/9315

**Reply by tiangolo** [reply]
Thank you for the fix @sharonyogev! :bow: 

This was just released in FastAPI version 0.95.1 :tada:

**Comment by JeroennC** [comment]
This issue also affects the non-annotated approach, probably limited to the decorator dependencies. Minimal example:
```python
from typing import Annotated

from fastapi import FastAPI, Header, Depends, APIRouter

app = FastAPI()


router = APIRouter()


def validate_header(x_test: Annotated[str | None, Header()] = None):
    if x_test != "test":
        raise Exception("Error")


@router.get("/", dependencies=[Depends(validate_header)])
async def root():
    return {"message": "Hello World"}


app.include_router(router)
```

From debugging it is caused by `analyze_param` being called more than once due to the `include_router`. `analyze_param` validates `field_info.default` in L387, but then may change it in L393. The second time `analyze_param` comes around is when the assertion on L387 raises an exception.

**Reply by tiangolo** [reply]
This should be fixed in https://github.com/tiangolo/fastapi/pull/9315

Just released in FastAPI 0.95.1 :tada:

**Comment by adriangb** [comment]
Could this be related to https://github.com/pydantic/pydantic/pull/4067? Maybe it’s worth trying to update Pydantic?

**Reply by sharonyogev** [reply]
I don't think so, take a look at my comment with my PR above, it's in FastAPI, but the bug and the solution is really similar

**Reply by adriangb** [reply]
thanks for confirming

**Comment by javierguzman** [comment]
Is there a workaround in the meantime the merge comes? I am having the same issue I believe, specifically I get this error:
```
Cookie` default value cannot be set in `Annotated` for 'authzAccessToken'. Set the default value with `=` instead.
```

Thank you in advance and regards

**Reply by tiangolo** [reply]
This should be fixed in https://github.com/tiangolo/fastapi/pull/9315

Just released in FastAPI 0.95.1 :tada:

**Comment by brianloss** [comment]
I'm seeing this as well. The docs provide an example that causes the error. Just try to run the example here: https://fastapi.tiangolo.com/tutorial/path-params-numeric-validations/#import-path

**Reply by tiangolo** [reply]
This should be fixed in https://github.com/tiangolo/fastapi/pull/9315

Just released in FastAPI 0.95.1 :tada:

**Comment by gitpushdashf** [comment]
Running into this as well. Definitely seems like a bug.

**Reply by tiangolo** [reply]
This should be fixed in https://github.com/tiangolo/fastapi/pull/9315

Just released in FastAPI 0.95.1 :tada:

**Comment by n8sty** [comment]
Can you all try your use cases with the fix from https://github.com/tiangolo/fastapi/discussions/9279#discussioncomment-5431795? If it works, confirm on this thread and if it doesn't note that too

**Reply by tlinhart** [reply]
I can confirm this fixes my original issue :+1:

**Reply by fnep** [reply]
Testing with 
```
pip install --force-reinstall git+https://github.com/sharonyogev/fastapi.git@fix/deepcoy_fieldinfo_of_annotated_args
```
… for me the issue is solved.

**Reply by javierguzman** [reply]
I did not know I could install from Github with pip...I am going to give a try as well thanks @fnep

**Reply by hochstibe** [reply]
The fix solves my issue

**Reply by bkis** [reply]
Tested it. Same here. #9315 would fix the issue for me, too.

**Reply by javierguzman** [reply]
Working for me as well

**Reply by tiangolo** [reply]
Thanks for the idea @n8sty! :nerd_face: :coffee: 

This should be fixed in https://github.com/tiangolo/fastapi/pull/9315

Just released in FastAPI 0.95.1 :tada:

**Comment by espdev** [comment]
I have the same problem in my project structure something like this:

```
app
  __init__.py
  api
      __init__.py
      routes.py
  main.py
```

```python
# routes.py

from typing import Annotated
from fastapi import APIRouter, Query

router = APIRouter()

@router.get('/hello')
async def hello(name: Annotated[str | None, Query(description='Your name')] = None):
    return f'Hello {name or "noname"}!'
```

```python
# main.py

from fastapi import FastAPI
from app import api

app = FastAPI()
app.include_router(api.router)
```

There is an exception:
```
AssertionError: `Query` default value cannot be set in `Annotated` for 'name'. Set the default value with `=` instead.
```

I tried running the code in the debugger and found out that the `analyze_param` function is called twice for the query parameter. 

The first call is correct, and it sets the default value for the query parameter:
![2023-04-07_12-03-01](https://user-images.githubusercontent.com/1299189/230579211-4a30dde9-ea2f-4f66-a1ce-df3dfcaf3e4f.png)

And the second call is failed because the default value is already set for the query parameter:
![2023-04-07_12-05-56](https://user-images.githubusercontent.com/1299189/230579714-7056aab3-3abb-4f22-a851-54ce2705ba0c.png)

I think this is a bug in FastAPI. Signatures and parameters should not be re-analyzed. This is non-idempotent behavior anyway.

**Reply by tiangolo** [reply]
This should be fixed in https://github.com/tiangolo/fastapi/pull/9315

Just released in FastAPI 0.95.1 :tada:

**Comment by mateoradman** [comment]
It's been really long now. Is this problem even considered an issue? The docs of v0.95.0 make it seem like this is definitely expected to work but it doesn't. It would be great if either the proposed solution #9315 gets reviewed and merged, or docs get updated explicitly stating this limitation of the "Annotated approach". In case of latter, please remove the tip `Prefer to use the Annotated version if possible.` which is shown in many sections of the documentation.

**Reply by amchii** [reply]
From this tweet (https://twitter.com/tiangolo/status/1645600824658378752?s=20)  , it can be seen that Tiangolo seems to be on vacation. However, I am curious whether only he has the permission to merge that PR in the entire FastAPI project.

**Reply by leokhachatorians** [reply]
As far as I know he's the only one with write access

**Reply by mateoradman** [reply]
With so many individuals and companies using and essentially relying on FastAPI nowadays, having a single contributor calling all the shots is suboptimal for a project of this size and importance.

**Reply by Kludex** [reply]
This problem is considered an issue. @tiangolo is already aware of it.

**Reply by mateoradman** [reply]
Thanks for letting us know @Kludex 🙂

**Reply by tiangolo** [reply]
This should be fixed in https://github.com/tiangolo/fastapi/pull/9315

Just released in FastAPI 0.95.1 :tada: 

@mateoradman your help would be much appreciated: https://fastapi.tiangolo.com/help-fastapi/

Actually, most of the work that needs to be done is not hitting the "merge" button, but reviewing PRs: https://fastapi.tiangolo.com/help-fastapi/#review-pull-requests

And answering questions: https://fastapi.tiangolo.com/help-fastapi/#help-others-with-questions-in-github

For the next time, you could go to the PR and add a review with a comment explaining how you tested the code and how it solved your problem. :nerd_face: That would be much appreciated and would help and speed up things a lot! :coffee:

**Comment by jdiaz-lw** [comment]
I have been bitten by this as well, worst of all is that I was preparing a POC to introduce FastAPI in my company so that we could adopt it across the whole Python teams :/ I think having the docs with those directions and not working like this can be a put off for tech leads assessing the adoption of the library.

**Reply by tiangolo** [reply]
This should be fixed in https://github.com/tiangolo/fastapi/pull/9315

Just released in FastAPI 0.95.1 :tada:

**Reply by mateoradman** [reply]
Thanks for merging it and letting us know, @tiangolo 🙂

**Comment by druidance** [comment]
Hey, 

It seems this still isn't working. Using this code snippet 
`v: Annotated[list[str], Query(default=[])]`
gives me 
`AssertionError: 'Query' default value cannot be set in 'Annotated' for 'v'. Set the default value with '=' instead.
`

I don't want to use current workaround
``v: Annotated[list[str], Query()] = []``
because I want to use Depends as a default value. But it also doesn't work.

Any suggestions how to solve this issue?

**FastAPI Version**
0.110.0

**Python Version**
Python 3.11.6

**Reply by sharonyogev** [reply]
For this I had this PR:
https://github.com/tiangolo/fastapi/pull/9389
But @tiangolo didn't want to merge it

**Reply by druidance** [reply]
Thank you for the PR. I will have to use another workaround then :-/

**Reply by Raphael2b3** [reply]
I am on v0.110.0

i need this to, i have routes, that share some query parameters, and i want to give them all the same default value, so it would be cool to

```
size_query = Annotated[int, Query(default=16,title="Size",description="Insert the number of random bytes to be encoded", ge=1, le=1000000)]
...
async def route(size:size_query): ...
etc..
```
